java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java

```diff
@@ -248,15 +238,17 @@ private AccessToken refreshAccessToken(int attempt) throws IOException {
                 }
 
                 tokenResponse = exchangeRefreshTokenForAccessToken();
-                this.currentToken = asAccessToken(tokenResponse);
-                storeSharedAccessTokenIfSupported(this.currentToken);
+                token = asAccessToken(tokenResponse);
+
+                storeSharedAccessTokenIfSupported(token);
 
                 if (lockNeeded) {
+                    //q: hold lock extra long, to wait for consistency??
```

Comment:
to limit risk of stale read, should we hold lock extra long here?

Reply:
I just found another issue here, in the previous lines:
```
                if (!acquired) {
                    //NOTE: check shared access token again, in case the instance which held the
                    // lock refreshed the token
                    Uninterruptibles.sleepUninterruptibly(attempt * WAIT_AFTER_FAILED_LOCK_ATTEMPTS, TimeUnit.MILLISECONDS);
 --------->       return refreshAccessToken(attempt + 1); 
                }
```
missing `return` in the recursive call, so might exchange again after a successful one or releasing a lock that might be in progress. Potential inconsistencies.

Said this, adding some random wait before releasing the lock is not a bad idea.